src/Security/RolesMatrixBuilder.php

```diff
@@ -0,0 +1,197 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\UserBundle\Security;
+
+use Sonata\AdminBundle\Admin\AdminInterface;
+use Sonata\AdminBundle\Admin\Pool;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * @author Silas Joisten <silasjoisten@hotmail.de>
+ */
+final class RolesMatrixBuilder implements RolesBuilderInterface
+{
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    /**
+     * @var AuthorizationCheckerInterface
+     */
+    private $authorizationChecker;
+
+    /**
+     * @var Pool
+     */
+    private $pool;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var string []
+     */
+    private $rolesHierarchy;
+
+    /**
+     * @var string []
+     */
+    private $excludeAdmin = [];
+
+    public function __construct(
+        TokenStorageInterface $tokenStorage,
+        AuthorizationCheckerInterface $authorizationChecker,
+        Pool $pool,
+        array $rolesHierarchy = []
+    ) {
+        $this->tokenStorage = $tokenStorage;
+        $this->authorizationChecker = $authorizationChecker;
+        $this->pool = $pool;
+        $this->rolesHierarchy = $rolesHierarchy;
+    }
+
+    public function setTranslator(TranslatorInterface $translator): void
+    {
+        $this->translator = $translator;
+    }
+
+    public function getRoles(string $domain = null, bool $expanded = true): array
+    {
+        if (!$this->tokenStorage->getToken()) {
+            return [];
+        }
+
+        $adminRoles = $this->getAdminRoles($domain);
+        $securityRoles = $this->getSecurityRoles($this->rolesHierarchy, $domain, $expanded);
+
+        return array_merge($securityRoles, $adminRoles);
```

Comment:
Please get rid of these OTVs:
```php
return array_merge(
    $this->getSecurityRoles($this->rolesHierarchy, $domain, $expanded),
    $this->getAdminRoles($domain)
);
```

Reply:
Both methods return the same kind of thing, since you merge them into one array. Please extract both methods into 2 classes, that will make it way easier to test for you.